lib/compiler/passes/report-duplicate-labels.js

```diff
@@ -36,9 +36,7 @@ function reportDuplicateLabels(ast) {
       const label = node.label;
       if (label && Object.prototype.hasOwnProperty.call(env, label)) {
         throw new GrammarError(
-          "Label \"" + node.label + "\" is already defined "
-            + "at line " + env[label].start.line + ", "
-            + "column " + env[label].start.column + ".",
```

Comment:
This is actually a useful information that should be kept. Some other messages just put there their own location, that is useless.

Ideally, this information also should be presented in the error object as a separate field, but right now I can't imagine which format would be better

Reply:
I agree that this should be reintroduced.  However, I also prefer the basic sentiment that the underlying message should be short.

I propose:

1. A "shortMessage" field be introduced
2. The full name be constructed by a function from the various fields
3. The following arguments be added to the constructor, then promoted as fields in the downstream product:
    1. start (object)
        1. line (number)
        2. column (number)
        3. offset (number)
    3. end (object)
        1. line (number)
        2. column (number)
        3. offset (number)
    4. length (number)

Reply:
This would imply that the string being passed to `GrammarError` is actually the `shortMessage`

Reply:
I'd be happy to do the work if we land this without this piece

Reply:
In a later commit, I added a `toString()`, the `codeLocation` bits that @Mingun wanted, and added a `referenceLocation` field for the errors that say "you have a problem *here* that references a thing *there*", e.g. duplicate labels.  I'm open to length, but not sure what value it adds.  From [pegjs-util](https://github.com/rse/pegjs-util#cooked-error-reporting), I really like `found`, and wouldn't mind the `prolog`, `token`, and `epilog` fields if they end up being helpful.

Reply:
> The following arguments be added to the constructor, then promoted as fields in the downstream product:

`GrammarError` already have a `location` that have more full information.

>  From [pegjs-util](https://github.com/rse/pegjs-util#cooked-error-reporting), I really like `found`, and wouldn't mind the `prolog`, `token`, and `epilog` fields if they end up being helpful.

`found` is a part of `SyntaxError` already and the mentioned method processes that kind of errors. I don't think that `prolog`, `token`, and `epilog` fields will be useful, but we can provide method in the `SyntaxError` that extracts any token from input stream and error position just by executing specified "errorTokenRule" rule:
```js
class SyntaxError extends Error {
  /// @param errorTokenRule?
  ///        Rule to extract logical token with error.
  ///        If not specified, grammar's default will be used
  ///        That default can be specified or in options,
  ///        or via annotations when annotations landed
  errorToken(input, errorTokenRule) {
    // Something like that
    return parser.parse(
      input.substr(this.location.start.offset),
      { startRule: errorTokenRule }
    );
  }
}
```